File: python/bindings/vqc_types.py
"""
API Bindings for the VQC types.
"""
from __future__ import annotations
from enum import Enum
from typing import Any

from bindings.vqc_dafny_core import EnsureDafnyCore
from bindings.vqc_currency import Money

EnsureDafnyCore()

import Orders as OrdersModule
import Types as TypesModule
import Validation as ValidationModule

Orders = OrdersModule.default__
Types = TypesModule
Validation = ValidationModule.default__

Account = Types.Account_Account
Ledger = Types.Ledger_Ledger
OrderRecord = Types.Order_Order
FillRecord = Types.Fill_Fill
PositionRecord = Types.Position_Position


class OrderStatus(Enum):
    """Broker-neutral lifecycle states supported by the verified core."""

    NEW = "new"
    ACCEPTED = "accepted"
    PARTIALLY_FILLED = "partially_filled"
    FILLED = "filled"
    CANCELLED = "cancelled"
    REJECTED = "rejected"


class OrderSide(Enum):
    """Economic direction of an order."""

    BUY = "buy"
    SELL = "sell"


class OrderType(Enum):
    """Primitive order instructions represented by VQC."""

    MARKET = "market"
    LIMIT = "limit"
    STOP = "stop"
    STOP_LIMIT = "stop_limit"


def Order(
    order_id: int,
    symbol: str,
    quantity: int,
    side: OrderSide = OrderSide.BUY,
    order_type: OrderType = OrderType.MARKET,
    status: OrderStatus = OrderStatus.NEW,
    filled_quantity: int = 0,
    limit_price: Money | None = None,
    stop_price: Money | None = None,
) -> OrderRecord:
    """Create a typed Dafny order record from the public Python enums."""
    if side is OrderSide.BUY:
        side_enum = Types.OrderSide_Buy()
    elif side is OrderSide.SELL:
        side_enum = Types.OrderSide_Sell()
    else:
        raise ValueError(f"unsupported side: {side}")

    if order_type is OrderType.MARKET:
        order_type_enum = Types.OrderType_Market()
    elif order_type is OrderType.LIMIT:
        if limit_price is None:
            raise ValueError("limit order requires a limit_price")
        order_type_enum = Types.OrderType_Limit(limit_price)
    elif order_type is OrderType.STOP:
        if stop_price is None:
            raise ValueError("stop order requires a stop_price")
        order_type_enum = Types.OrderType_Stop(stop_price)
    elif order_type is OrderType.STOP_LIMIT:
        if stop_price is None or limit_price is None:
            raise ValueError("stop-limit order requires stop_price and limit_price")
        order_type_enum = Types.OrderType_StopLimit(stop_price, limit_price)
    else:
        raise ValueError(f"unsupported order type: {order_type}")

    status_enum = _to_dafny_order_status(status)

    return OrderRecord(
        order_id,
        symbol,
        quantity,
        side_enum,
        order_type_enum,
        status_enum,
        filled_quantity,
    )


def Fill(
    execution_id: int,
    order_id: int,
    symbol: str,
    quantity: int,
    price: Money,
    timestamp: int,
) -> FillRecord:
    """Create a priced execution record for an existing order."""
    return FillRecord(execution_id, order_id, symbol, quantity, price, timestamp)


def Position(symbol: str, quantity: int, average_price: Money) -> PositionRecord:
    """Create a long-position record using whole-share quantity."""
    return PositionRecord(symbol, quantity, average_price)


def _to_dafny_order_status(status: OrderStatus) -> Any:
    """Translate the public enum into its generated Dafny counterpart."""
    if status is OrderStatus.NEW:
        return Types.OrderStatus_New()
    if status is OrderStatus.ACCEPTED:
        return Types.OrderStatus_Accepted()
    if status is OrderStatus.PARTIALLY_FILLED:
        return Types.OrderStatus_PartiallyFilled()
    if status is OrderStatus.FILLED:
        return Types.OrderStatus_Filled()
    if status is OrderStatus.CANCELLED:
        return Types.OrderStatus_Cancelled()
    if status is OrderStatus.REJECTED:
        return Types.OrderStatus_Rejected()
    raise ValueError(f"unsupported status: {status}")


RemainingQuantity = Orders.RemainingQuantity


IsValidOrder = Validation.IsValidOrder
IsValidFill = Validation.IsValidFill
IsValidPosition = Validation.IsValidPosition
IsValidLedger = Validation.IsValidLedger
IsValidAccount = Validation.IsValidAccount

File: typescript/bindings/vqc_types.ts
/** Bindings for VQC record construction, order utility, and validation. */

import { Dafny, IntegerInput, Natural, StringValue } from "./vqc_dafny_core";
import { Money } from "./vqc_currency";

declare const accountBrand: unique symbol;
declare const ledgerBrand: unique symbol;
declare const orderBrand: unique symbol;
declare const fillBrand: unique symbol;
declare const positionBrand: unique symbol;

export type Account = { readonly [accountBrand]: "Account" };
export type Ledger = { readonly [ledgerBrand]: "Ledger" };
export type OrderRecord = { readonly [orderBrand]: "Order" };
export type FillRecord = { readonly [fillBrand]: "Fill" };
export type PositionRecord = { readonly [positionBrand]: "Position" };

export type OrderSide = "buy" | "sell";
export type OrderType = "market" | "limit" | "stop" | "stop_limit";
export type OrderStatus =
    | "pending"
    | "open"
    | "partially_filled"
    | "filled"
    | "cancelled"
    | "rejected";

type DafnyOrderSide = { readonly _dafnyOrderSide: unique symbol };
type DafnyOrderType = { readonly _dafnyOrderType: unique symbol };
type DafnyOrderStatus = { readonly _dafnyOrderStatus: unique symbol };

function toDafnyOrderSide(side: OrderSide): DafnyOrderSide {
    if (side === "buy") {
        return Dafny.Types.OrderSide.create_Buy() as DafnyOrderSide;
    }
    if (side === "sell") {
        return Dafny.Types.OrderSide.create_Sell() as DafnyOrderSide;
    }
    throw new Error(`unsupported side: ${side}`);
}

/** @internal Converts a public status to the generated Dafny status value. */
export function toDafnyOrderStatus(status: OrderStatus): DafnyOrderStatus {
    switch (status) {
        case "pending": return Dafny.Types.OrderStatus.create_Pending() as DafnyOrderStatus;
        case "open": return Dafny.Types.OrderStatus.create_Open() as DafnyOrderStatus;
        case "partially_filled": return Dafny.Types.OrderStatus.create_PartiallyFilled() as DafnyOrderStatus;
        case "filled": return Dafny.Types.OrderStatus.create_Filled() as DafnyOrderStatus;
        case "cancelled": return Dafny.Types.OrderStatus.create_Cancelled() as DafnyOrderStatus;
        case "rejected": return Dafny.Types.OrderStatus.create_Rejected() as DafnyOrderStatus;
    }
}

/** Creates a VQC order record. */
export function Order(
    orderId: IntegerInput,
    symbol: string,
    quantity: IntegerInput,
    side: OrderSide = "buy",
    orderType: OrderType = "market",
    status: OrderStatus = "pending",
    filledQuantity: IntegerInput = 0,
    limitPrice?: Money,
    stopPrice?: Money,
): OrderRecord {
    let orderTypeEnum: DafnyOrderType;
    if (orderType === "market") {
        orderTypeEnum = Dafny.Types.OrderType.create_Market() as DafnyOrderType;
    } else if (orderType === "limit") {
        if (limitPrice === undefined) {
            throw new Error("limit order requires a limitPrice");
        }
        orderTypeEnum = Dafny.Types.OrderType.create_Limit(limitPrice) as DafnyOrderType;
    } else if (orderType === "stop") {
        if (stopPrice === undefined) {
            throw new Error("stop order requires a stopPrice");
        }
        orderTypeEnum = Dafny.Types.OrderType.create_Stop(stopPrice) as DafnyOrderType;
    } else if (orderType === "stop_limit") {
        if (stopPrice === undefined || limitPrice === undefined) {
            throw new Error("stop-limit order requires stopPrice and limitPrice");
        }
        orderTypeEnum = Dafny.Types.OrderType.create_StopLimit(
            stopPrice,
            limitPrice,
        ) as DafnyOrderType;
    } else {
        throw new Error(`unsupported order type: ${orderType}`);
    }

    return Dafny.Types.Order.create_Order(
        Natural(orderId, "orderId"),
        StringValue(symbol, "symbol"),
        Natural(quantity, "quantity"),
        toDafnyOrderSide(side),
        orderTypeEnum,
        toDafnyOrderStatus(status),
        Natural(filledQuantity, "filledQuantity"),
    ) as OrderRecord;
}

/** Creates a VQC fill record for an existing order. */
export function Fill(
    executionId: IntegerInput,
    orderId: IntegerInput,
    symbol: string,
    quantity: IntegerInput,
    price: Money,
    timestamp: IntegerInput,
): FillRecord {
    return Dafny.Types.Fill.create_Fill(
        Natural(executionId, "executionId"),
        Natural(orderId, "orderId"),
        StringValue(symbol, "symbol"),
        Natural(quantity, "quantity"),
        price,
        Natural(timestamp, "timestamp"),
    ) as FillRecord;
}

/** Creates a VQC long-position record. */
export function Position(
    symbol: string,
    quantity: IntegerInput,
    averagePrice: Money,
): PositionRecord {
    return Dafny.Types.Position.create_Position(
        StringValue(symbol, "symbol"),
        Natural(quantity, "quantity"),
        averagePrice,
    ) as PositionRecord;
}

export function RemainingQuantity(order: OrderRecord): string {
    return Dafny.Orders.__default.RemainingQuantity(order).toFixed();
}

export function IsValidOrder(order: OrderRecord): boolean {
    return Dafny.Validation.__default.IsValidOrder(order);
}

export function IsValidFill(fill: FillRecord): boolean {
    return Dafny.Validation.__default.IsValidFill(fill);
}

export function IsValidPosition(position: PositionRecord): boolean {
    return Dafny.Validation.__default.IsValidPosition(position);
}

export function IsValidLedger(ledger: Ledger): boolean {
    return Dafny.Validation.__default.IsValidLedger(ledger);
}

export function IsValidAccount(account: Account): boolean {
    return Dafny.Validation.__default.IsValidAccount(account);
}

export const HaltException = Dafny._dafny.HaltException;
